File: C#Lox/Scanner.cs
using System.Globalization;

namespace Lox;

public class Scanner
{
    private static readonly Dictionary<string, TokenType> keywords = new()
    {
        { "and",    TokenType.AND    },
        { "class",  TokenType.CLASS  },
        { "else",   TokenType.ELSE   },
        { "false",  TokenType.FALSE  },
        { "for",    TokenType.FOR    },
        { "fun",    TokenType.FUN    },
        { "if",     TokenType.IF     },
        { "nil",    TokenType.NIL    },
        { "or",     TokenType.OR     },
        { "print",  TokenType.PRINT  },
        { "return", TokenType.RETURN },
        { "super",  TokenType.SUPER  },
        { "this",   TokenType.THIS   },
        { "true",   TokenType.TRUE   },
        { "var",    TokenType.VAR    },
        { "while",  TokenType.WHILE  },
        { "break",  TokenType.BREAK  }
    };
    
    private readonly string source;
    private readonly List<Token> tokens = [];

    private int start = 0;
    private int current = 0;
    private int line = 1;

    public Scanner(string source)
    {
        this.source = source;
    }

    public async Task<List<Token>> ScanTokensAsync()
    {
        while (!IsAtEnd())
        {
            start = current;
            await ScanTokenAsync();
        }

        tokens.Add(new Token(TokenType.EOF, string.Empty, null, line));
        return tokens;
    }

    private async Task ScanTokenAsync()
    {
        char c = Advance();

        switch (c)
        {
            case '(':
                AddToken(TokenType.LEFT_PAREN);
                break;

            case ')':
                AddToken(TokenType.RIGHT_PAREN);
                break;

            case '{':
                AddToken(TokenType.LEFT_BRACE);
                break;

            case '}':
                AddToken(TokenType.RIGHT_BRACE);
                break;

            case ',':
                AddToken(TokenType.COMMA);
                break;

            case '.':
                AddToken(TokenType.DOT);
                break;

            case '-':
                AddToken(TokenType.MINUS);
                break;

            case '+':
                AddToken(TokenType.PLUS);
                break;

            case ';':
                AddToken(TokenType.SEMICOLON);
                break;

            case '*':
                AddToken(TokenType.STAR);
                break;

            case '?':
                AddToken(TokenType.QUESTION_MARK);
                break;

            case ':':
                AddToken(TokenType.COLON);
                break;

            case '!':
                AddToken(Match('=') ? TokenType.BANG_EQUAL : TokenType.BANG);
                break;

            case '=':
                AddToken(Match('=') ? TokenType.EQUAL_EQUAL : TokenType.EQUAL);
                break;

            case '<':
                AddToken(Match('=') ? TokenType.LESS_EQUAL : TokenType.LESS);
                break;

            case '>':
                AddToken(Match('=') ? TokenType.GREATER_EQUAL : TokenType.GREATER);
                break;

            case '/':
                if (Match('/'))
                {
                    while (Peek() != '\n' && !IsAtEnd())
                    {
                        _ = Advance();
                    }
                }
                else if (Match('*'))
                {
                    while (Peek() != '*' && PeekNext() != '/')
                    {
                        if (Advance() == '\n')
                        {
                            line++;
                        }
                    }

                    // Discard '*' and '/'
                    _ = Advance();
                    _ = Advance();
                }
                else
                {
                    AddToken(TokenType.SLASH);
                }
                break;

            case ' ':
            case '\r':
            case '\t':
                // Ignore
                break;

            case '\n':
                line++;
                break;

            case '"':
                await ScanStringAsync();
                break;

            default:
                if (IsDigit(c))
                {
                    ScanNumber();
                }
                else if (IsAlpha(c))
                {
                    ScanIdentifier();
                }
                else
                {
                    await Lox.ErrorAsync(line, "Unexpected character.");
                }
                break;
        }
    }

    private void ScanIdentifier()
    {
        while (IsAlphanumeric(Peek()))
        {
            _ = Advance();
        }

        string text = source[start..current];
        
        if (!keywords.TryGetValue(text, out TokenType type))
        {
            type = TokenType.IDENTIFIER;
        }

        AddToken(type);
    }

    private void ScanNumber()
    {
        while (IsDigit(Peek()))
        {
            _ = Advance();
        }

        // Look for a fractional part
        if (Peek() == '.' && IsDigit(PeekNext()))
        {
            _ = Advance(); // consume the '.'

            while (IsDigit(Peek()))
            {
                _ = Advance();
            }
        }

        var doubleLiteral = double.Parse(source[start..current], CultureInfo.InvariantCulture);

        AddToken(TokenType.NUMBER, doubleLiteral);
    }

    private static bool IsDigit(char c)
    {
        return c is >= '0' and <= '9';
    }

    private async Task ScanStringAsync()
    {
        while (Peek() != '"' && !IsAtEnd())
        {
            if (Peek() == '\n')
            {
                line++;
            }

            _ = Advance();
        }

        if (IsAtEnd())
        {
            await Lox.ErrorAsync(line, "Unterminated string.");
            return;
        }

        _ = Advance(); // Closing '"'

        var stringLiteral = source[(start + 1)..(current - 1)];

        AddToken(TokenType.STRING, stringLiteral);
    }

    private char Peek()
    {
        return IsAtEnd() ? '\0' : source[current];
    }

    private char PeekNext()
    {
        if (current + 1 >= source.Length)
        {
            return '\0';
        }
        else
        {
            return source[current + 1];
        }
    }

    private static bool IsAlpha(char c)
    {
        return (c is >= 'a' and <= 'z')
            || (c is >= 'A' and <= 'Z')
            || c == '_';
    }

    private static bool IsAlphanumeric(char c)
    {
        return IsAlpha(c) || IsDigit(c);
    }

    // Advance and return true if the next character from source equals expected.
    // Return false otherwise.
    private bool Match(char expected)
    {
        if (IsAtEnd() || source[current] != expected)
        {
            return false;
        }
        else
        {
            current++;
            return true;
        }
    }

    private bool IsAtEnd()
    {
        return current >= source.Length;
    }

    private char Advance()
    {
        return source[current++];
    }

    private void AddToken(TokenType type)
    {
        AddToken(type, null);
    }

    private void AddToken(TokenType type, object? literal)
    {
        var text = source[start..current];
        tokens.Add(new Token(type, text, literal, line));
    }
}

File: C++Lox/scanner.cpp

#include "scanner.hpp"

bool lox::scanner::is_at_end()
{
    return m_current == m_source.length();
}

lox::token lox::scanner::make_token(const lox::token_type type)
{
    return
    {
        type,
        m_source.substr(m_start, m_current - m_start),
        m_line
    };
}

lox::token lox::scanner::error_token(const std::string_view message)
{
    return
    {
        token_type::ERROR,
        message,
        m_line
    };
}

char lox::scanner::advance()
{
    m_current++;

    return m_source[m_current - 1];
}

bool lox::scanner::match(char expected)
{
    if (is_at_end()) return false;

    if (m_source[m_current] != expected) return false;

    m_current++;

    return true;
}

char lox::scanner::peek()
{
    if (is_at_end()) return '\0';

    return m_source[m_current];
}

char lox::scanner::peek_next()
{
    if (is_at_end() || m_current + 1 == m_source.length()) return '\0';

    return m_source[m_current + 1];
}

void lox::scanner::skip_whitespace()
{
    for (;;)
    {
        char c = peek();

        switch (c)
        {
        case ' ':
        case '\r':
        case '\t':
            advance();
            break;
        case '\n':
            m_line++;
            advance();
            break;
        case '/':
            if (peek_next() == '/')
            {
                // A comment goes until the end of the line.
                while (peek() != '\n' && !is_at_end()) advance();
            }
            else
            {
                return;
            }
            break;
        default:
            return;
        }
    }
}

lox::token lox::scanner::make_string()
{
    while (peek() != '"' && !is_at_end())
    {
        if (peek() == '\n') m_line++;

        advance();
    }

    if (is_at_end()) return error_token("Unterminated string.");

    // The closing quote.
    advance();

    return make_token(token_type::STRING);
}

bool lox::scanner::is_digit(char c)
{
    return c >= '0' && c <= '9';
}

lox::token lox::scanner::number()
{
    while (is_digit(peek())) advance();

    // Look for a fractional part.
    if (peek() == '.' && is_digit(peek_next()))
    {
        // Consume the ".".
        advance();

        while (is_digit(peek())) advance();
    }

    return make_token(token_type::NUMBER);
}

bool lox::scanner::is_alpha(char c)
{
    return (c >= 'a' && c <= 'z') ||
        (c >= 'A' && c <= 'Z') ||
        c == '_';
}

lox::token_type lox::scanner::check_keyword(const std::size_t start, const std::string_view rest, const token_type type)
{
    if (m_current - m_start == start + rest.length()
        && rest.compare(m_source.substr(m_start + start, rest.length())) == 0)
    {
        return type;
    }

    return token_type::IDENTIFIER;
}

lox::token_type lox::scanner::identifier_type()
{
    switch (m_source[m_start])
    {
    case 'a': return check_keyword(1, "nd", token_type::AND);
    case 'c': return check_keyword(1, "lass", token_type::CLASS);
    case 'e': return check_keyword(1, "lse", token_type::ELSE);
    case 'f':
        if (m_current - m_start > 1)
        {
            switch (m_source[m_start + 1])
            {
            case 'a': return check_keyword(2, "lse", token_type::FALSE);
            case 'o': return check_keyword(2, "r", token_type::FOR);
            case 'u': return check_keyword(2, "n", token_type::FUN);
            }
        }
        break;
    case 'i': return check_keyword(1, "f", token_type::IF);
    case 'n': return check_keyword(1, "il", token_type::NIL);
    case 'o': return check_keyword(1, "r", token_type::OR);
    case 'p': return check_keyword(1, "rint", token_type::PRINT);
    case 'r': return check_keyword(1, "eturn", token_type::RETURN);
    case 's': return check_keyword(1, "uper", token_type::SUPER);
    case 't':
        if (m_current - m_start > 1)
        {
            switch (m_source[m_start + 1])
            {
            case 'h': return check_keyword(2, "is", token_type::THIS);
            case 'r': return check_keyword(2, "ue", token_type::TRUE);
            }
        }
        break;
    case 'v': return check_keyword(1, "ar", token_type::VAR);
    case 'w': return check_keyword(1, "hile", token_type::WHILE);
    }

    return token_type::IDENTIFIER;
}

lox::token lox::scanner::identifier()
{
    while (is_alpha(peek()) || is_digit(peek())) advance();

    return make_token(identifier_type());
}

lox::scanner::scanner(const std::string_view source) : m_source{ source }
{
    m_start = m_current = 0;
    m_line = 1;
}

lox::token lox::scanner::scan_token()
{
    skip_whitespace();

    m_start = m_current;

    if (is_at_end()) return make_token(token_type::END_OF_FILE);

    char c = advance();

    if (is_alpha(c)) return identifier();

    if (is_digit(c)) return number();

    switch (c)
    {
    case '(': return make_token(token_type::LEFT_PAREN);
    case ')': return make_token(token_type::RIGHT_PAREN);
    case '{': return make_token(token_type::LEFT_BRACE);
    case '}': return make_token(token_type::RIGHT_BRACE);
    case ';': return make_token(token_type::SEMICOLON);
    case ',': return make_token(token_type::COMMA);
    case '.': return make_token(token_type::DOT);
    case '-': return make_token(token_type::MINUS);
    case '+': return make_token(token_type::PLUS);
    case '/': return make_token(token_type::SLASH);
    case '*': return make_token(token_type::STAR);
    case '!': return make_token(match('=') ? token_type::BANG_EQUAL : token_type::BANG);
    case '=': return make_token(match('=') ? token_type::EQUAL_EQUAL : token_type::EQUAL);
    case '<': return make_token(match('=') ? token_type::LESS_EQUAL : token_type::LESS);
    case '>': return make_token(match('=') ? token_type::GREATER_EQUAL : token_type::GREATER);
    case '"': return make_string();
    }

    return error_token("Unexpected character.");
}
